debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorTask.java

```diff
@@ -151,12 +153,28 @@ public ChangeEventSourceCoordinator start(Configuration config) {
                     .loggingContextSupplier(() -> taskContext.configureLoggingContext(CONTEXT_NAME))
                     .build();
 
-            errorHandler = new PostgresErrorHandler(connectorConfig.getLogicalName(), queue);
+            ErrorHandler errorHandler = new PostgresErrorHandler(connectorConfig.getLogicalName(), queue);
 
             final PostgresEventMetadataProvider metadataProvider = new PostgresEventMetadataProvider();
 
-            Heartbeat heartbeat = Heartbeat.create(connectorConfig.getConfig(), topicSelector.getHeartbeatTopic(),
-                    connectorConfig.getLogicalName(), heartbeatConnection);
+            Configuration configuration = connectorConfig.getConfig();
+            Heartbeat heartbeat = Heartbeat.create(
+                    configuration.getDuration(Heartbeat.HEARTBEAT_INTERVAL, ChronoUnit.MILLIS),
+                    configuration.getString(DatabaseHeartbeatImpl.HEARTBEAT_ACTION_QUERY),
+                    topicSelector.getHeartbeatTopic(),
+                    connectorConfig.getLogicalName(), heartbeatConnection, exception -> {
+                        String sqlErrorId = exception.getSQLState();
+                        switch (sqlErrorId) {
+                            case "57P01":
+                                // Postgres error admin_shutdown, see https://www.postgresql.org/docs/12/errcodes-appendix.html
+                                throw new DebeziumException("Could not execute heartbeat action (Error: " + sqlErrorId + ")", exception);
+                            case "57P03":
+                                // Postgres error cannot_connect_now, see https://www.postgresql.org/docs/12/errcodes-appendix.html
+                                throw new RetriableException("Could not execute heartbeat action (Error: " + sqlErrorId + ")", exception);
```

Comment:
@jpechane I don't think that this is a reason for a graceful shutdown of the connector. I think the connector should go into FAILED state and the user needs to resolve the Postgres issue. Then the user can simply resume the connector via Kafka Connect REST API.
`cannot_connect_now` can also maybe also happen during boot or reboot and might be retriable (depending on the timing). At least we can now specify the behavior per connector and maybe see how it behaves in the wild, wdyt?

Reply:
@rk3rn3r You are right!